concordium-consensus/src/Concordium/GlobalState/Persistent/BlobStore.hs

```diff
@@ -521,73 +636,60 @@ class Monad m => Reference m ref a where
   refUncache :: ref a -> m (ref a)
 
 -- |A value that may exists purely on disk ('BRBlobbed'), purely in memory
--- ('BRMemory' with @brIORef = null@), or both in memory and on disk. When the
+-- ('BRMemory' with @brIORef = Null@), or both in memory and on disk. When the
 -- value is both on disk and in memory the two values must match.
 data BufferedRef a
     = BRBlobbed {brRef :: !(BlobRef a)}
     -- ^Value stored on disk
-    | BRMemory {brIORef :: !(IORef (BlobRef a)), brValue :: !a}
+    | BRMemory {brIORef :: !(IORef (Nullable (BufferedRef a))), brValue :: !a}
     -- ^Value stored in memory and possibly on disk.
-    -- When a new 'BRMemory' instance is created, we initialize 'brIORef' to 'refNull'.
-    -- When we store the instance in persistent storage, we update 'brIORef' with the corresponding pointer.
+    -- When a new 'BRMemory' instance is created, we initialize 'brIORef' to 'Null'.
+    -- When we store the instance in persistent storage, we update 'brIORef' to a @Some BRBoth{..}@.
     -- That way, when we store the same instance again on disk (this could be, e.g., a child block
     -- that inherited its parent's state) we can store the pointer to the 'brValue' data rather than
     -- storing all of the data again.
+    -- This definition makes it possible for shared references to share the
+    -- updated value after it's been 'storeUpdated'.
+    -- See implementation for 'storeUpdate'.
```

Comment:
This comment does not tell me (maybe it should, but I am not getting it) why, for example, I would not use 
```
    | BRMemory {brIORef :: !(IORef (Nullable (!a, !BlobRef a))), brValue :: !a}
```
seeing that the only variant that is ever put in is BRBoth (I realize the pair above should be strict and you can't put bangs on nested values, but you get my point). What would be wrong with this?

But also the fact that now BRMemory is not really that anymore is confusing. It sounds like there is some insight here that was gained through a lot of effort, but is not documented.

Please describe a specific example where sharing this IORef between multiple BRMemory values provides an improvement. Describe it in such a way that one can actually observe it in the future when any changes to this are needed.